test_data/asm/keccak.asm

```diff
@@ -0,0 +1,200 @@
+
+// ln 42 - 94
+// compression function
+let keccakf: int[] -> int[] = |st| swap_u64_loop(keccakf_inner(swap_u64_loop(st))); // int[25] -> int[25]
+
+// ln 96 - 141
+// TODO: to_bytes and from_bytes are implemented below but I'm not sure if we have existing helper functions to use
+// needs to be little endian
+let to_bytes: int[] -> int[] = |input| // int[25] -> int[200] 
+    array::fold(input, [], |acc, elem| {
+        let new_bytes = array::new(8, |i| 
+            // elem % (1 << 64) / (1 << 56)
+            // elem % (1 << 56) / (1 << 48)
+            // elem % (1 << 48) / (1 << 40)
+            // ...
+            elem % (1 << (64 - 8 * i)) / (1 << (56 - 8 * i))
+        );
+        acc + new_bytes
+    });
+
+// needs to be little endian
+let from_bytes: int[] -> int[] = |input| // int[200] -> int[25]
+    array::new(25, |i| 
+        utils::fold(8, |j| j, 0, |acc, idx| 
+            acc + input[i * 8 + 7 - idx] * (1 << (8 * idx))
+        )
+    );
+
+// ln 148 - 158
+// int[N], int, int -> int[25]
+let update_finalize_b: int[], int, int -> int[] = |input, rate, delim| {
+    let num_loop = array::len(input) / rate;
+    let num_remaining = array::len(input) % rate;
+    let b_delim_idx = (num_remaining + 1) % rate;
+
+    let b_keccak = utils::fold(num_loop, |i| i, array::new(200, |i| 0), |acc, idx| {
+        let new_b = array::zip(array::new(rate, |i| acc[i]), array::new(rate, |i| input[idx * rate + i]), xor);
+        let new_b_pad = array::new(200, |i| if i < rate { new_b[i] } else { acc[i] });
+        to_bytes(keccakf(from_bytes(new_b_pad)))
+    });
+    let b_finalize = array::new(200, |i| 
+        // num_remaining is 0 the minimum and rate - 1 the maximum
+        if i < num_remaining {
+            // ln 150, one of the remaining to be xor'ed
+            xor(b_keccak[i], input[num_loop * rate + i])
+        } else {
+            if i == num_remaining {
+                if i == rate - 1 {
+                    xor_mult([b_keccak[i], delim, 0x80])
+                } else {
+                    xor(b_keccak[i], delim)
+                }
+            } else {
+                if i == rate - 1 {
+                    xor(b_keccak[i], 0x80)
+                } else {
+                    b_keccak[i]
+                }
+            }
+        }
+    );
+    // ln 158
+    to_bytes(keccakf(from_bytes(b_finalize)))    
+};
+
+// ln 143 - 161
+// W is output number of bytes, input is array of bytes, delim is a single byte
+let main: int, int[], int -> int[] = |W, input, delim| { 
+    // ln 144 - 145
+    let rate = 200 - (2 * W); // int, 100 if u32
+
+    let b_finalized = update_finalize_b(input,rate, delim); // 200 bytes
+
+    // ln 160
+    // note that the biggest W is 64, i.e. 512-bit keccak
+    // therefore, W should always be lower than byte length of b_finalized (200)
+    array::new(W, |i| b_finalized[i])
+};
+
+// main machine
+// TODO: not sure how to integrate these
```

Comment:
Please make sure that there are no TODOs left (at least those that are in scope here)

Reply:
Done